src/algebra/homology/chain_complex.lean

```diff
@@ -0,0 +1,104 @@
+import category_theory.graded_object
+import category_theory.differential_object
+
+universes v u
+
+open category_theory
+open category_theory.limits
+
+section
+variables (V : Type u) [𝒱 : category.{v} V]
+include 𝒱
+
+variables [has_zero_morphisms.{v} V]
+
+/--
+A chain complex in `V` is "just" a differential `ℤ`-graded object in `V`.
+-/
+-- Actually, because I chose the "shift" to correspond to the +1 direction,
+-- this is what would usually be called a cochain complex.
+-- We should definitely bikeshed this. :-)
+@[nolint has_inhabited_instance]
+def chain_complex : Type (max v u) :=
+differential_object.{v} (graded_object ℤ V)
+
+-- The chain groups of a chain complex `C` are accessed as `C.X i`,
+-- and the differentials as `C.d i : C.X i ⟶ C.X (i+1)`.
+example (C : chain_complex V) : C.X 5 ⟶ C.X 6 := C.d 5
+
+variables {V}
+/--
+A convenience lemma for morphisms of differential graded objects,
+picking out one component of the commutation relation.
+-/
+-- Could this be a simp lemma? Which way?
+lemma category_theory.differential_object.hom.comm_at {X Y : differential_object.{v} (graded_object ℤ V)}
+  (f : X ⟶ Y) (i : ℤ) : f.f i ≫ Y.d i = X.d i ≫ f.f (i+1) := congr_fun f.comm i
+end
+
+namespace chain_complex
+variables {V : Type u} [𝒱 : category.{v} V]
+include 𝒱
+
+variables [has_zero_morphisms.{v} V]
+
+@[simp]
+lemma d_squared (C : chain_complex.{v} V) (i : ℤ) :
+  C.d i ≫ C.d (i+1) = 0 :=
+congr_fun (C.d_squared) i
+
+-- FIXME why doesn't this work `by simp`?
+example (C : chain_complex V) : C.d 5 ≫ C.d 6 = 0 := C.d_squared 5
```

Comment:
You probably need simp-lemmas in terms of `bit0` and `bit1`.
(I think) `simp` doesn't recognise that `6 = 2*3` is `5+1 = (2*2+1) + 1`.

Reply:
Ugh, I see. Writing a lemma in terms of bit0 and bit1, you can cover the `C.d even \gg C.d odd` cases, but dependent type theory hell prevents you doing the other cases. I think I'm going to remove this example for now. :-)